### PR TITLE
feat: create password reset route

### PR DESCRIPTION
**Pull request by Myles181**
## Title
Implement Password Reset Functionality via Email OTP

## Description
Implemented a password reset feature using FastAPI. This feature allows users to request a password reset by providing their email address. An OTP code is generated and sent to the user's email address, which they can use to reset their password. The endpoint does not change the password directly; it only handles OTP generation and email sending.

## Related Issue
This pull request is related to the following issue: [287](https://github.com/hngprojects/hng_boilerplate_python_fastapi_web/issues/297)

## Motivation and Context
This change is required to provide a secure and user-friendly way for users to reset their passwords in case they forget it. It enhances the user experience by allowing them to receive an OTP code via email, ensuring that only the rightful owner of the email address can reset the password.

## How Has This Been Tested?
- The endpoint was tested using Postman to ensure that it correctly sends an OTP to the provided email address.
- Various scenarios were tested, including:
  - Valid email address
  - Non-existent email address
  - Missing email field
- The OTP generation and email sending functionalities were tested to ensure they work as expected.
- The database was checked to ensure OTP records are created with correct expiration times.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Additional Notes
- The `OTP` model was created to store OTP codes with fields for email, OTP code, creation time, and expiration time.
- The `password_reset` endpoint was added to handle the password reset requests.
- Custom exception handling was implemented to manage various error scenarios.

### CODE REVIEW

**Review by joboy-dev — CHANGES_REQUESTED** [review]
Why didn't you make use of services to handle DB queries??

File: api/v1/models/password_reset.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+from api.v1.models.base_model import BaseTableModel
+from datetime import datetime, timedelta
+
+
+class OTP(BaseTableModel):
+    __tablename__ = "otps"
+
+    email = Column(String, index=True, nullable=False)
+    otp_code = Column(Integer, nullable=False)
+    created_at = Column(DateTime, default=datetime.utcnow())
```
Comment (joboy-dev) [inline api/v1/models/password_reset.py:13]: Remove this created_at

File: api/v1/schemas/password_reset.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from pydantic import BaseModel, EmailStr
+
+
+class EMAILSCHEMA(BaseModel):
```
Comment (joboy-dev) [inline api/v1/schemas/password_reset.py:4]: change to EmailSchema

**Review by joboy-dev — CHANGES_REQUESTED** [review]
Look at suggestions

File: api/v1/models/password_reset.py (hunk 1 above)
Comment (Myles181) [reply api/v1/models/password_reset.py:13]: alright

Comment (Myles181) [reply api/v1/models/password_reset.py:13]: I've removed it sir

File: api/v1/schemas/password_reset.py (hunk 2 above)
Comment (Myles181) [reply api/v1/schemas/password_reset.py:4]: Done sir

File: api/v1/models/newsletter.py [hunk 3]
```diff
@@ -17,4 +17,4 @@ class Newsletter(BaseTableModel):
     title = Column(String, nullable=False)
     content = Column(Text, nullable=False)
 
-    subscribers = relationship("User", secondary=user_newsletter_association, back_populates="newsletters")
\ No newline at end of file
+    subscribers = relationship("User", secondary=user_newsletter_association, back_populates="newsletters")
```
Comment (Goketech) [inline api/v1/models/newsletter.py:20]: revert this change

**Review by Goketech — CHANGES_REQUESTED** [review]

**Pull request closed without merging**